packages/flutter_tools/lib/src/mdns_discovery.dart

```diff
@@ -149,25 +179,33 @@ class MDnsObservatoryDiscovery {
     bool usesIpv6 = false,
     int? hostVmservicePort,
     int? deviceVmservicePort,
+    bool isNetworkDevice = false,
   }) async {
     final MDnsObservatoryDiscoveryResult? result = await query(
       applicationId: applicationId,
       deviceVmservicePort: deviceVmservicePort,
+      ipv6: usesIpv6,
+      isNetworkDevice: isNetworkDevice,
     );
     if (result == null) {
       await _checkForIPv4LinkLocal(device);
       return null;
     }
-
-    final String host = usesIpv6
+    final String host;
+    if (isNetworkDevice && result.ipAddress != null) {
+      host = result.ipAddress!.address;
```

Comment:
You can avoid the `!` by [type promotion](https://dart.dev/tools/non-promotion-reasons#property-or-this) of the local.

```suggestion
    final String? ipAddress = result.ipAddress;
    if (isNetworkDevice && ipAddress != null) {
      host = ipAddress.address;
```